sys/kern/subr_intr.c

```diff
@@ -460,12 +460,11 @@ isrc_free_irq(struct intr_irqsrc *isrc)
 	mtx_assert(&isrc_table_lock, MA_OWNED);
 
 	if (isrc->isrc_irq >= intr_nirq)
-		return (EINVAL);
-	if (irq_sources[isrc->isrc_irq] != isrc)
-		return (EINVAL);
+		return (0);
 
-	irq_sources[isrc->isrc_irq] = NULL;
-	isrc->isrc_irq = INTR_IRQ_INVALID;	/* just to be safe */
+	if (irq_sources[isrc->isrc_irq] == isrc)
+		irq_sources[isrc->isrc_irq] = NULL;
+	isrc->isrc_irq = INTR_IRQ_INVALID;	/* mark removed */
 
```

Comment:
We shouldn't be papering over programing errors.  It's a programming error when isrc_irq is >= intr_nirq. That normally can't happen. Likewise with setting things to NULL in this way. This is a bad change. NAK.

Hard to say for sure, though, since there's not actually any callers of this in the tree. This is only called from `intr_isrc_deregister`, which is global, but has no actual callers. So how did you test this?

But this is unsafe. If they don't match, it must be an error.

Reply:
> We shouldn't be papering over programing errors. It's a programming error when isrc_irq is >= intr_nirq. That normally can't happen. Likewise with setting things to NULL in this way. This is a bad change. NAK.
A hypothetical scenario, but one which I believe _should_ be handled as it could nominally be used.  Imagine a highly hot-plug bus on a system with >1000 processors.  On such a hot-plug bus it could nominally be valid to have one thread which calls `intr_isrc_deregister()` at 1 second intervals and a separate thread doing `intr_teardown_irq()` calls.  In such a case the sequence:
```
intr_isrc_deregister();
intr_teardown_irq();
intr_isrc_deregister();
```
Issue is the first `intr_isrc_deregister()` call would get `EBUSY` from `intr_event_destroy()` since the filter wasn't yet removed.  Presently the second would get `EINVAL` from `isrc_free_irq()` since the `isrc` had already been removed from the table by the first call.  When it comes down to it, either `isrc_free_irq()` or `isrc_event_destroy()` **must** tolerate being called multiple times with the same `isrc`.

> Hard to say for sure, though, since there's not actually any callers of this in the tree. This is only called from `intr_isrc_deregister`, which is global, but has no actual callers. So how did you test this?
By reading the source and concluding this was a nominally valid course of action.  I've got a driver which is expected to call `intr_isrc_deregister()` during normal operation.  As such I'm examining `intr_isrc_deregister()` and discovering it is rather buggy.  The x86 equivalent driver is unable to release interrupts and is thus forced to keep them on a list.  Whereas INTRNG has the function partially implemented, but untested.

I have no expectation to need this functionality, but when trying to fix the problems with `intr_isrc_deregister()` this issue shows up.  Presently I'm unable to submit the driver since 3 key areas it needs working aren't in usable shape.